text/0000-coroutines.md

```diff
@@ -0,0 +1,485 @@
+
+# Alternatives
+
+## Alternatives to stackless coroutines
+
+### Stackful coroutines
+
+Stackful coroutines are amenable to library implementation and, in the first approximation, do not require language-level support. 
+Indeed, [crates.io](https://crates.io/search?q=coroutine) already contains at least half a dozen of such crates.
+They also come with some drawbacks:
+- Each needs a  stack of at least a few kilobytes in size, which makes them much costlier than needed (our example above
+  would have needed only a few dozen bytes for coroutine state). 
+- Yielding and resuming requires a register context swap, which is relatively slow.
+- Switching stacks is fragile: operating systems often make the assumption that they are the only ones managing 
+  thread's register context. 
+
+### Source transformations
+
+There are precedents in other languages of implementing similar functionality via purely source transformations. 
+One such example is [F#'s "Computation Expressions"](http://msdn.microsoft.com/en-us/library/dd233182.aspx).  
+
+The exact F# approach is unlikely to work in Rust because of Rust's strong ownership semantics.  F# workflows transform 
+control flow constructs into a tree of nested closures, which share variables with the parent closure.  In Rust such sharing 
+is prohibited when variables are mutable.
+
+Another plausible approach is a source-to-source transformation into an state machine, where states are represented 
+as variants of an enum, each containing hoisted variables that are live during transition to that state. 
+An example of such approach is [Stateful](https://github.com/erickt/stateful) 
+([blog post](http://erickt.github.io/blog/2016/01/27/stateful-in-progress-generators/)).  
+The difficulty here is in computing liveless and types of local variables.  The former is needed to decide 
+which variables need to be hoisted, the latter - to declare their types in the "State" enum.
+It seems that an implementation would either need to replicate may Rust compiler passes (name resolution,
+type inference, possibly borrow checking), or find some clever way to leverage the compiler to compute this 
+information.  In the end, it is probably much easier to implement such transformation as a MIR transform, 
+which is what this RFC is about.
+
+## Alternatives designs
+
+### Implicitly binding coroutine arguments
+Under the current design, the values of parameters with which a coroutine is resumed, are returned as a tuple
+from the **yield** expression.
+This looks a bit awkward, especially for single-argument coroutines, where it would be cleaner to return the argument 
+itself, instead of a 1-tuple wrapping it:
+```rust
+|a1| {
+    let (b1,) = yield x;
+}
+```
+An alternative would be to implicitly bind coroutine arguments to the same names after each yield point:
+```rust
+|a1| {
+    yield x;  // actually means let (a1,) = yield x; 
+}
```

Comment:
Is there no potential third alternative here of adding some sort of special case for single-argument coroutines that has a chance of working reasonably well?